### DISCUSSION TITLE
Extracting DBC file from DB

### DISCUSSION THREAD

**Discussion by roizano** [cantools/cantools#502] [Q&A]
Hi,
I would like to know if there is any way to extract DBC file out from DB.
In general, I have DBC with many messages and I would like to split it into several files according to the messages.
Any idea if CANTOOLS can support this?

**Comment by juleq** [comment]
Hi,

you can load the dbc, e.g. as db1.
You can create a new db, e.g. as db2.
You can deep copy the wanted messages from db1 to db2.
You can dump db2 into a dbc.

It might be worth calling the refresh function to rebuild indices used for easy navigation on db2.

Check your dbc for completeness, e.g. are all attributes you want in it or should you add a copy of those, too?

Cannot go into more detail, but I think you can do it like this.

**Reply by roizano** [reply]
@juleq  Thank you for the reply.
I loaded the db1.
how to create a new db2 and hard copy db2?
I managed to copy the message and all other properties but the dump_file failed all the time on :
TypeError: as_dbc_string() missing 1 required positional argument: 'self'

**Reply by juleq** [reply]
Good job. Maybe I can spot it when you post the failing line and 2-3 lines before it.

**Reply by roizano** [reply]
@juleq 
This is what I used:
########
db = cantools.database.load_file("koko.dbc")  ##Loading DBC
db2 = db.get_message_by_frame_id(int(5))     ##Loading specific message number5
# create new dbc
db_final = type(db) 
db_final.messages = db2
db_final.autosar= autosar
db_final.buses = bus
db_final.nodes = nodes
db_final.dbc =  dbc
db_final.version = version
db_final.refresh
## dump new dbc
cantools.database.dump_file(db_final, 'bar.dbc')
######################

Thank you :)

**Reply by roizano** [reply]
Hi @juleq 
I managed to work on that :) 
by removing the old messages and assign new message :)

**Reply by juleq** [reply]
Good to hear. Cantools is really useful to eleminate a lot of dbc twiddeling footwork. I also appreciate it for that and just plug it into the build system.
